src/core.rs

```diff
@@ -0,0 +1,247 @@
+        use pretty_assertions::assert_eq;
+
+        #[test]
+        fn digits() {
+            let expected = (0..=9).collect::<Vec<_>>();
+            let result = (b'0'..=b'9')
+                .map(|c| parse_hex_digit(c).unwrap())
+                .collect::<Vec<_>>();
+
+            assert_eq!(result, expected);
+        }
+
+        #[test]
+        fn lower() {
+            let expected = (10..=15).collect::<Vec<_>>();
+            let result = (b'a'..=b'f')
+                .map(|c| parse_hex_digit(c as u8).unwrap())
+                .collect::<Vec<_>>();
+
+            assert_eq!(result, expected);
+        }
+
+        #[test]
+        fn upper() {
+            let expected = (10..=15).collect::<Vec<_>>();
+            let result = (b'A'..=b'F')
+                .map(|c| parse_hex_digit(c as u8).unwrap())
+                .collect::<Vec<_>>();
+
+            assert_eq!(result, expected);
+        }
+
+        #[test]
+        fn nonhex() {
+            assert_eq!(
+                parse_hex_digit(b'g'),
+                Err(HexParseError::InvalidNibble('g'))
+            );
+            assert_eq!(
+                parse_hex_digit(b'z'),
+                Err(HexParseError::InvalidNibble('z'))
+            );
+            assert_eq!(
+                parse_hex_digit(b'0' - 1),
+                Err(HexParseError::InvalidNibble((b'0' - 1) as char))
+            );
+            assert_eq!(
+                parse_hex_digit(b'A' - 1),
+                Err(HexParseError::InvalidNibble((b'A' - 1) as char))
+            );
+        }
+    }
+
+    mod stark_hash {
+        use super::*;
+
+        mod serde {
+            use crate::core::StarkHash;
+
+            #[test]
+            fn round_trip() {
+                let expected = [
+                    0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23, 0x45, 0x67, 0x89,
+                    0xAB, 0xCD, 0xEF, 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23,
+                    0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF,
+                ];
+
+                let hash = StarkHash::from_be_bytes(expected);
+                let bytes = hash.to_be_bytes();
+                assert_eq!(bytes, expected);
+            }
+        }
+
+        mod from_hex_str {
+            use super::*;
+            use pretty_assertions::assert_eq;
+
+            /// Tests that [StarkHash::from_hex_str] returns the expected result.
+            ///
+            /// Repeats this test both with and without a "0x" prefix.
+            fn run_test_case(test_str: &str, expected: Result<StarkHash, HexParseError>) {
+                let without_prefix = StarkHash::from_hex_str(test_str);
+                assert_eq!(without_prefix, expected);
+
+                let with_prefix = StarkHash::from_hex_str(&format!("0x{}", test_str));
+                assert_eq!(with_prefix, expected);
+            }
+
+            #[test]
+            fn empty() {
+                run_test_case("", Ok(StarkHash([0; 32])));
```

Comment:
I'd argue that this is an invalid hash and we should have an error variant for this case, ie.: `"0x"` and `""`.